tests/csv_datums/test_csv_sources.py

```diff
@@ -90,35 +106,157 @@ def source_interval():
-    assert requests.received_get == make_get("https://api.kraken.com/0/public/Trades", pair="xbtusd",
-                                             since=1559347200000000000)
+@pytest.fixture(autouse=True)
+def local_time(monkeypatch, source_interval):
+    s = LocalTimeSpy()
+    monkeypatch.setattr(module_under_test, 'time', s)
+    return s
 
 
-def make_get(url, **params):
-    return GetRequest(url, params)
+@pytest.fixture
+def server_time(monkeypatch, source_interval):
+    s = ServerTimeStub()
+    s.set_current_time(1559347200 + source_interval + 1)
+    monkeypatch.setattr(module_under_test, 'KrakenServerTime', lambda: s)
+    return s
 
 
-def test_kraken_source_returns_validated_and_adapted_data(source, source_interval, requests):
-    requests.set_get_response(json={'some': 'data'})
-    adapted = Adapted(json={'some': 'data'}, with_interval=source_interval)
-    assert source.query(since=1559347200) == Validated(csv_datums_from(adapted))
+@pytest.fixture
+def make_json(server_time, source_interval):
+    def factory(results, last=None):
+        last = last or to_nano_sec(server_time.current_time + source_interval + 1)
+        results['last'] = str(last)
+        return {'result': results}
+
+    return factory
+
+
+def make_get(url, **params):
+    return GetRequest(url, params)
 
 
 def csv_datums_from(adapted):
     return CsvDatums(adapted.interval, adapted)
+
+
+@pytest.mark.usefixtures("server_time")
+class TestKrakenSource:
+    def test_kraken_source_builds_correct_url(self, source, requests):
+        source.query(since=1559347200)
+        assert requests.received_get == make_get("https://api.kraken.com/0/public/Trades", pair="xbtusd",
+                                                 since=1559347200000000000)
+
+    def test_kraken_source_returns_validated_and_adapted_data(self, source, source_interval, requests, validation,
+                                                              make_json):
+        requests.set_get_response(json=make_json({'pair': [['data']]}))
+        adapted = AdaptedData(json=make_json({'pair': [['data']]}), with_interval=source_interval)
+        assert source.query(since=1559347200) == csv_datums_from(adapted)
+        assert validation.data == csv_datums_from(adapted)
+
+    def test_kraken_source_queries_data_up_to_last_incomplete_interval(self, source, source_interval, requests,
+                                                                       server_time, make_json):
+        server_time.set_current_time(1559347200 + source_interval * 3 * 60 + 10)
+        requests.set_get_responses(
+            make_json({'pair': [1]}, last=to_nano_sec(1559347200 + source_interval * 60)),
+            make_json({'pair': [2]}, last=to_nano_sec(1559347200 + source_interval * 1.5 * 60)),
+            make_json({'pair': [3]}, last=to_nano_sec(1559347200 + source_interval * 3 * 60)),
+            make_json({'pair': [4]}, last=to_nano_sec(1559347200 + source_interval * 3 * 60 + 1)),
```

Comment:
should be refactored into a make_pair or make_trades factory function in case we need to change this format